src/box/lua/config/applier/credentials.lua

```diff
@@ -653,6 +653,12 @@ local function sync_privileges(credentials, obj_to_sync)
     end)
 end
 
+box.ctl.on_schema_upgrade_end(function()
+    if not config then return end
+
+    sync_privileges(get_credentials(config))
+end)
```

Comment:
It is a partial solution. We don't create users and roles, don't set passwords after an upgrade.

AFAIU, the right solution would be replacing the RW checks (`not box.info.ro`) with some 'DDL is enabled' checks (RW + not needs upgrade) and replacing RW waiting with 'DDL is enabled' waiting. We can use `on_replace` on `_schema` for the latter.

Also, ignoring of `box.error.SCHEMA_NEEDS_UPGRADE` is not needed with a full solution of the problem.

Reply:
Thank you. Now it's clearer how to fix it.
Applied the suggested way to fix the problem with `on_replace`.
Removed ignoring `SCHEMA_NEEDS_UPGRADE`